ext/RMagick/rmagick.h

```diff
@@ -995,7 +1031,11 @@ extern VALUE  TextureFill_fill(VALUE, VALUE);
 ATTR_ACCESSOR(Pixel, red)
 ATTR_ACCESSOR(Pixel, green)
 ATTR_ACCESSOR(Pixel, blue)
+#if defined(IMAGEMAGICK_7)
+ATTR_ACCESSOR(Pixel, alpha)
```

Comment:
I'm a little vague on what these `ATTR_ACCESSOR` calls are doing. Are they the same thing as a bare `attr_accessor` in ruby? If so, can we alias the two? Or maybe we can override the `opacity` method and add a deprecation warning to it, letting users know to use `alpha`.

Reply:
Alpha and opacity are two different things. For alpha 255 (Q8) is fully opaque and zero is transparent. But for opacity this is the other way around, zero is opaque and 255 is transparent. In ImageMagick 6 there is no alpha and if you would add this you would always need to convert this value. This is also one of the reasons for me that there should be two different API's and two different versions (or gems) for ImageMagick 6 and 7.

Reply:
If it's not too hard to convert, then I'd advocate for us to do that. I think it would be ideal if we can maintain existing RMagick functionality across both IM6 and IM7. I think it's fine to add new functionality only for IM7, but I kind of want to dig in on each of these changes and see if there's a way to preserve existing behavior for the time being. I know that's going to be more work in the short term, but I think it's going to save us a lot in the long term.

Some of the benefits I see of trying to maintain existing functionality:
- We can provide a smooth upgrade path for our users with deprecation warnings. This is probably the biggest one for me. Having gone through the pain of upgrading a major version on packages where there are huge changes, I feel pretty strongly about making this process as painless as possible for our users and allowing them to update their code gradually without a big bang package upgrade. I really admire libraries like [RSpec](https://relishapp.com/rspec/docs/upgrade) for making upgrades really easy. Basically, if you've updated to the most recent minor version and fixed all of the deprecations, then bumping to the next major version should cause no breakage. I know this is much harder for a tool like RMagick that depends on another library like ImageMagick which has changed so much, but I think it's going to be worth the effort if we can pull it off. I've also seen [other libraries](https://github.com/rails-api/active_model_serializers#status-of-ams) that are in perpetual version confusion due to major breaking changes between versions.
- We avoid user confusion with which version to install. If they're looking just at Rubygems, they might just install the latest version regardless, and then they'll report issues against RMagick 7.x when they're actually using IM6. We'll need extra documentation and code checks to prevent these issues. It'll also avoid us needing to answer questions about which version to use, since it's a less conventional way of handling versioning.
- We avoid needing to manage shipping two versions simultaneously. Right now the deploy is based on the checked in `RMagick::VERSION`, so we'd need additional logic to support this, or we'd need two branches running alongside each other.
- It sounds like a fun challenge. I know I haven't contributed a whole lot of code up to this point. You and @Watson1978 have been doing an excellent job, so I've been content to cheer from the sidelines, but I'm happy to help however I can. I love thinking through these kinds of problems, so I can help find a solution if you can help me understand the problem. I can also try to write some C where it might be helpful, but I'll need your expertise to make sure I don't introduce horrendous bugs.

Reply:
- I would advise against doing a conversion. ImageMagick 7 uses alpha throughout the whole code but ImageMagick 6 uses opacity. If you really want to pretend that you are using alpha in ImageMagick 6 you will probably need a lot of hacks and extra complexity. I have also done the same transition with the C# library of ImageMagick when going from version 6 to 7. Having a clear distinct difference between the old and the new version will make users more careful when upgrading. And they should be careful because a lot of the internals have changed.
- I share your concern that users might be confused what to install. While writing this response I got an extra idea. We could also decide to create two new gems. One called rmagick6 and one called rmagick7 to make it more clear which version of ImageMagick should be used. But having a gem called rmagick6 and rmagick will also make a clear distinction between both of them. And even a version number could make that clear. I do think that the key factor is communication here. If we explain this in the readme on github and in the details of the gem users will know which gem/version to use.
- I don't mind adding additional logic/scripts to make it easy to distribute two version of RMagick. With the strategy from this PR we keep both versions very close to each other and we only need to make the split in the end when the gem is distributed.
- Would love having you help with the `C` part and you will definitely learn a lot from it. I do have a lot of code cleanup ideas were we could really use some extra help. And I love to do reviews, we will both learn from them.

Reply:
>     * ImageMagick 7 uses alpha throughout the whole code but ImageMagick 6

Could we set up methods to make this easier? For example something like this (in pseudocode):

```
set_alpha(image, value) {
  if (IM_7) {
    image->alpha = value;
  } else {
    image->opacity = 255 - value;
  }
}

set_opacity(image, value) {
  rb_warn("deprecated...");
  set_alpha(255 - value);
}
```

Reply:
It's not just the RMagick code but the whole internal code of ImageMagick 7 now uses alpha instead of opacity. If you are doing operations with for example the `fx` method on the image And I am not saying you can't, just saying that you shouldn't add these kind of hacks. And in the future we should really change some of the pixel API's because there are a lot more things that can be done in IM7 due to the new channel layout. That should be something for the future but that will never ever work in IM6 and then we would end up with those "not supported" methods that I really dislike.

Reply:
Here's another random idea to avoid the "not supported" problem. What if we have `RMagick` proper be cross compatible, but then we have another gem, something like `RMagick-IM7-extensions` which adds functionality for IM7?

Reply:
My first instinct says that I would prefer to keep the RMagick module work with the most recent version of ImageMagick without having to install something extra. And that will only solve what has been added. Things that are no longer supported will still raise a "not supported" error.

What would you think about doing a big step in version number and go to 7.0.0 for RMagick linked with IM7 and a new gem called RMagick6 that is linked with IM6? Maybe this will also make it more clear that something big has happened when we suddenly jump from 3.x to 7?

Reply:
> p.s. I like that you are challenging my ideas and coming up with new ideas on how to handle this.

Ha, I'm glad you're willing to continue discussing this!

I don't think we'd need to bump the version to 7.0 to indicate that there are major changes. 4.0 should be enough. It'll make things tricky, too, if we make breaking changes that are unrelated to IM7. Do we then bump it to 8.0?

Thinking more about the idea of having multiple gems, though. What if we have an `RMagick`, `RMagick7` and an `RMagick-core` which has shared logic for both of them? Since most of our users are likely to be still using ImageMagick 6 at this point, I think it makes sense to keep that as the primary version, and then for people who are eager to jump on IM7, they can change their gem. Once we're ready to deprecate IM6 support, we can merge `RMagick7` into `RMagick` proper. We could alternatively have `RMagick6` and `RMagick7` gems, but again depending on a common `RMagick-core`.

The above seems workable, though I still don't love it. I think the ideal situation is that we have a single `RMagick` gem that smartly handles the transition between ImageMagick versions. Otherwise, our users have to update their RMagick version, their code, and their ImageMagick version all at once on their system, which just sounds like a painful process. Another option would be to simply keep an ImageMagick7 branch on the repo here that people can make use of, and hopefully over time we can figure out how to smooth the transition more.

One more idea: how crazy would it be to have an ImageMagick 6.10 release that makes this transition easier? Maybe it could add some of the interfaces coming up in IM7.

I think it would be helpful for us to generate a table of the RMagick interfaces that are going to change, how things worked before and what users will need to change them to after.

Reply:
After a chat with @mockdeep we came to the conclusion that we could change the `opacity` property to `alpha` first and negate the channel for ImageMagick 6. This should be moved to a new PR.